src/ol/Image.js

```diff
@@ -142,39 +142,47 @@ class ImageWrapper extends ImageBase {
    * @private
    */
   unlistenImage_() {
-    unlistenImage(this.imageListenerKeys_);
+    if (this.unlisten_) {
+      this.unlisten_();
+      this.unlisten_ = null;
+    }
   }
 }
 
 /**
  * @param {HTMLCanvasElement|HTMLImageElement|HTMLVideoElement} image Image element.
  * @param {function():any} loadHandler Load callback function.
  * @param {function():any} errorHandler Error callback function.
- * @return {Array<import("./events.js").EventsKey>} listener keys.
+ * @return {function():void} Callback to stop listening.
  */
 export function listenImage(image, loadHandler, errorHandler) {
   const img = /** @type {HTMLImageElement} */ (image);
   if (img.decode) {
-    img.decode().then(loadHandler).catch(errorHandler);
-    return null;
-  } else {
-    const listenerKeys = [
-      listenOnce(img, EventType.LOAD, loadHandler),
-      listenOnce(img, EventType.ERROR, errorHandler)
-    ];
-    return listenerKeys;
+    const promise = img.decode();
+    let listening = true;
+    const unlisten = function() {
+      listening = false;
+    };
+    promise.then(function() {
+      if (listening) {
+        loadHandler();
+      }
+    }).catch(function(error) {
+      if (listening) {
+        errorHandler();
+      }
+    });
```

Comment:
We should add a `finally` and call `unlisten` to make sure we have the same behavior as the `listenOnce` function.